File: PDOO/p4/Civitas/src/civitas/Tablero.java

package civitas;
import java.util.ArrayList;


public class Tablero {
    
    private int numCasillaCarcel;  
    private int porSalida;
    private Boolean tieneJuez;
    
    private ArrayList<Casilla> casillas;
    
        
    Tablero(int indiceCarcel){
        if (indiceCarcel >= 1){
            numCasillaCarcel = indiceCarcel;
        } else {
            numCasillaCarcel = 1;
        }
        
        casillas = new ArrayList<>();
        Casilla Salida = new Casilla("Salida");
        casillas.add(Salida);
        
        porSalida = 0;
        
        tieneJuez = false;
    }
    
    private Boolean correcto(){
        Boolean resultado = false;
        if (casillas.size() > numCasillaCarcel && tieneJuez == true){
            resultado = true;            
        }
        return resultado;
    }
    
    private Boolean correcto(int numCasilla){
        Boolean resultado = false;
        if (correcto() == true && numCasilla < casillas.size()){
            resultado = true;
        }
        return resultado;
    }
    
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        int resultado = porSalida;
        if (porSalida > 0){
            porSalida--;
        }
        return resultado;
    }
    
    void añadeCasilla(Casilla casilla){
        if (casillas.size() == numCasillaCarcel){
            Casilla Carcel = new Casilla("Carcel");
            casillas.add(Carcel);
        }
        if (casilla instanceof CasillaJuez){
            añadeJuez();
        } else {
            casillas.add(casilla);
        }
        if (casillas.size() == numCasillaCarcel){
            Casilla Carcel = new Casilla("Carcel");
            casillas.add(Carcel);
        }
    }
    
    void añadeJuez(){
        if(tieneJuez == false){
            CasillaJuez Juez = new CasillaJuez(numCasillaCarcel,"Juez");
            casillas.add(Juez);
            tieneJuez = true;
        }
    }
    
    Casilla getCasilla(int numCasilla){
        Casilla resultado = null;
        if (correcto(numCasilla) == true){
            resultado = casillas.get(numCasilla);
        }
        return resultado;
    }
    
    int nuevaPosicion(int actual, int tirada){
        int resultado;
        if (correcto() == false){
            resultado = -1;
        } else {
            resultado = actual + tirada;
            if (correcto(resultado) == false){
                int aux = resultado - casillas.size();
                resultado = aux;
            }
        }
        if (resultado != actual+tirada){
            porSalida++;
        }
        return resultado;
    }
    
    int calcularTirada(int origen, int destino){
        int resultado;
        resultado = destino - origen;
        if (resultado < 0){
            resultado = resultado + casillas.size();
        }
        return resultado;
    }
    
}

File: PDOO/p4/civitas_rb/lib/tablero.rb
require_relative 'casilla'


module Civitas

class Tablero
  def initialize (indiceCarcel)
    if indiceCarcel>=1
      @numCasillaCarcel=indiceCarcel
    else
      @numCasillaCarcel=1
    end
    
    @casillas = Array.new
    salida = Casilla.new("Salida")
    @casillas.push(salida)
     
    @porSalida=0
    @tieneJuez = false
  end
  
  def correcto
    resultado=false
    resultado=true if (@casillas.length() > @numCasillaCarcel && @tieneJuez == true)
    return resultado
  end
  
  def correcto2(numCasilla)
    resultad=false
    if correcto() == true && numCasilla < @casillas.length()
      resultad=true
    end
    return resultad
  end
  
  def get_por_salida
    resultado=@porSalida
    if(@porSalida>0)
      @porSalida=@porSalida-1
    end
    return resultado
  end
  
  def aniade_casilla(casilla)
    if @casillas.length== @numCasillaCarcel
      carcel=Casilla.new("Carcel")
      @casillas.push(carcel)
    end
    if casilla.instance_of? Casilla_juez
      aniade_juez()
    else
      @casillas.push(casilla)
    end
    if @casillas.length== @numCasillaCarcel
      carcel=Casilla.new("Carcel")
      @casillas.push(carcel)
    end
  end
  
  def aniade_juez
    if @tieneJuez==false
      juez = Casilla_juez.new(numCasillaCarcel, "Juez")
      @casillas.push(juez)
      @tieneJuez=true
    end
  end
  
  def get_casilla(numCasilla)
    return @casillas.at(numCasilla)
  end
  
  def nueva_posicion(actual, tirada)
    if correcto()==false
      resultado=-1
    else
      resultado=actual+tirada
      if correcto2(resultado)==false
        aux=resultado- @casillas.length
        resultado=aux
      end
    end
    if resultado != actual+tirada
      @porSalida=@porSalida+1
    end
    return resultado
  end
  
  def calcular_tirada(origen, destino)
    resultado=destino-origen 
    if resultado<0
      resultado=resultado +@casillas.length
    end
    return resultado
  end
  
  attr_accessor :casillas, :salida, :numCasillaCarcel, :porSalida
  
  private :correcto
end

end